// rn-swipeable-panel/lib/components/Close.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Close = void 0;
const React = require("react");
const react_native_1 = require("react-native");
exports.Close = ({ onPress, rootStyle, iconStyle }) => {
    return (<react_native_1.TouchableOpacity activeOpacity={0.5} onPress={() => onPress()} style={[CloseStyles.closeButton, rootStyle]}>
      <react_native_1.View style={[CloseStyles.iconLine, iconStyle, { transform: [{ rotateZ: '45deg' }] }]}/>
      <react_native_1.View style={[CloseStyles.iconLine, iconStyle, { transform: [{ rotateZ: '135deg' }] }]}/>
    </react_native_1.TouchableOpacity>);
};
const CloseStyles = react_native_1.StyleSheet.create({
    closeButton: {
        width: 30,
        height: 30,
        borderRadius: 15,
        position: 'absolute',
        right: 15,
        top: 15,
        backgroundColor: '#e2e2e2',
        zIndex: 3,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
    },
    iconLine: {
        position: 'absolute',
        width: 18,
        height: 2,
        borderRadius: 2,
        backgroundColor: 'white',
    },
});
